In [1]:
#Importing required Libraries
from pyspark import SparkContext # Spark
from pyspark.sql import SparkSession # Spark SQL
from pyspark.rdd import RDD
import pyspark.sql.functions as f
# Import SparkConf class into program
from pyspark import SparkConf


#  1.1
## Task 1

In [2]:
#Task1.1
#1

master = "local[*]"
# The `appName` field is a name to be shown on the Spark cluster UI page
app_name = "Assignment-1"
# Setup configuration parameters for Spark
spark_conf = SparkConf().setMaster(master).setAppName(app_name)

# Method 1: Using SparkSession
spark = SparkSession.builder.config(conf=spark_conf).getOrCreate()
sc = spark.sparkContext
sc.setLogLevel('ERROR')


#  1.1
## Task 2

In [3]:
#Task1.1
#2
# loading all the csv in one variable
unit_files='2015_DATA_SA_Units.csv,2016_DATA_SA_Units.csv,2017_DATA_SA_Units.csv,2018_DATA_SA_Units.csv,2019_DATA_SA_Units.csv'
# Creating an RDD for unit csv for 2015-16
unit_rdd = sc.textFile(unit_files)

# 1.1
## Task 3

In [4]:
#Task 1.1
#3
#loading all crash csv in a variable
crash_files='2015_DATA_SA_Crash.csv,2016_DATA_SA_Crash.csv,2017_DATA_SA_Crash.csv,2018_DATA_SA_Crash.csv,2019_DATA_SA_Crash.csv'
# loading all the csv in rdd form for 2015-16
crash_rdd = sc.textFile(crash_files) 

# 1.1
## Task4

In [5]:
#task1.1
#4

#Spliting each line in unit separated by comma into a list 
unit_rdd1 = unit_rdd.map(lambda line: line.split(','))

#filter out unit header
header_unit = unit_rdd1.first()

# Removing the header of unit rdd
unit_rdd1 = unit_rdd1.filter(lambda row: row != header_unit)   


# Spliting each line separated by comma into a list 
crash_rdd1 = crash_rdd.map(lambda line: line.split(','))

# filter out crash header
header_crash = crash_rdd1.first()

#Removing the header crash rdd
crash_rdd1 = crash_rdd1.filter(lambda row: row != header_crash)   


In [6]:
# Displaying count of crash RDD
crash_rdd1.count()

72006

In [7]:
# Displaying first 10 rows of crash RDD
crash_rdd1.take(10)

[['"2015-1-21/08/2019"',
  '"2 Metropolitan"',
  '"ELIZABETH VALE"',
  '"5112"',
  '"CITY OF PLAYFORD."',
  '2',
  '1',
  '0',
  '0',
  '1',
  '2015',
  '"January"',
  '"Wednesday"',
  '"01:00 pm"',
  '"060"',
  '"T-Junction"',
  '"Straight road"',
  '"Level"',
  '"Not Applicable"',
  '"Sealed"',
  '"Dry"',
  '"Not Raining"',
  '"Daylight"',
  '"Side Swipe"',
  '"01"',
  '"Driver Rider"',
  '"2: MI"',
  '"No Control"',
  '""',
  '""',
  '1335254.54',
  '1690056.88',
  '"13352551690057"'],
 ['"2015-2-21/08/2019"',
  '"2 Metropolitan"',
  '"SALISBURY"',
  '"5108"',
  '"CITY OF SALISBURY"',
  '2',
  '1',
  '0',
  '0',
  '1',
  '2015',
  '"February"',
  '"Tuesday"',
  '"03:38 pm"',
  '"060"',
  '"Cross Road"',
  '"Straight road"',
  '"Level"',
  '"Not Applicable"',
  '"Sealed"',
  '"Dry"',
  '"Not Raining"',
  '"Daylight"',
  '"Rear End"',
  '"01"',
  '"Driver Rider"',
  '"2: MI"',
  '"Traffic Signals"',
  '""',
  '""',
  '1333389.6',
  '1688248.34',
  '"13333901688248"'],
 ['"2015-3-21/08

In [8]:
# Displaying count of crash RDD
unit_rdd1.count()

153854

In [9]:
# Displaying first 10 rows of crash RDD
unit_rdd1.take(10)

[['"2015-1-21/08/2019"',
  '"01"',
  '0',
  '"SA"',
  '"RIGID TRUCK LGE GE 4.5T"',
  '"1999"',
  '"North East"',
  '"Male"',
  '"052"',
  '"SA"',
  '"HRR "',
  '"Full"',
  '"Not Towing"',
  '"Swerving"',
  '"001"',
  '"5109"',
  '',
  ''],
 ['"2015-1-21/08/2019"',
  '"02"',
  '1',
  '"SA"',
  '"Motor Cars - Sedan"',
  '"2009"',
  '"North East"',
  '"Female"',
  '"057"',
  '"SA"',
  '"C "',
  '"Full"',
  '"Not Towing"',
  '"Straight Ahead"',
  '"002"',
  '"5125"',
  '',
  ''],
 ['"2015-2-21/08/2019"',
  '"01"',
  '0',
  '"SA"',
  '"Motor Cars - Sedan"',
  '"2009"',
  '"South East"',
  '"Male"',
  '"020"',
  '"SA"',
  '"MR"',
  '"Provisional 1 "',
  '"Not Towing"',
  '"Straight Ahead"',
  '"001"',
  '"5110"',
  '',
  ''],
 ['"2015-2-21/08/2019"',
  '"02"',
  '1',
  '"SA"',
  '"Motor Cars - Sedan"',
  '"1994"',
  '"South East"',
  '"Female"',
  '"021"',
  '"SA"',
  '"C "',
  '"Full"',
  '"Not Towing"',
  '"Stopped on Carriageway"',
  '"001"',
  '"5096"',
  '',
  ''],
 ['"2015-3-21/08/2019

# 1.2
## Task 1

$ \text {There are 5 partition in each RDD for unit and crash data. In RDD when we explicitly do not specify any partitioning  }$

$\text {technique Rdd uses default method to partition the data this method is known as Round Robin partitioning strategy. } \$

In [10]:
#Task1.2
#1
#printing default partition in crash and unit Rdd
print('Default partitions in units: ',unit_rdd1.getNumPartitions())
print('Default partitions in crashes: ',crash_rdd1.getNumPartitions())

Default partitions in units:  5
Default partitions in crashes:  5


# 2.3
## Task 2A

In [11]:
#Task1.2
#2a
#Creating key value with key as the Lic state and others as value
flatMapValues_unit_rdd_2 = unit_rdd1.map(lambda x: (x[9], x[0:9]+x[10:]))
# displaying all the data
flatMapValues_unit_rdd_2.collect()

[('"SA"',
  ['"2015-1-21/08/2019"',
   '"01"',
   '0',
   '"SA"',
   '"RIGID TRUCK LGE GE 4.5T"',
   '"1999"',
   '"North East"',
   '"Male"',
   '"052"',
   '"HRR "',
   '"Full"',
   '"Not Towing"',
   '"Swerving"',
   '"001"',
   '"5109"',
   '',
   '']),
 ('"SA"',
  ['"2015-1-21/08/2019"',
   '"02"',
   '1',
   '"SA"',
   '"Motor Cars - Sedan"',
   '"2009"',
   '"North East"',
   '"Female"',
   '"057"',
   '"C "',
   '"Full"',
   '"Not Towing"',
   '"Straight Ahead"',
   '"002"',
   '"5125"',
   '',
   '']),
 ('"SA"',
  ['"2015-2-21/08/2019"',
   '"01"',
   '0',
   '"SA"',
   '"Motor Cars - Sedan"',
   '"2009"',
   '"South East"',
   '"Male"',
   '"020"',
   '"MR"',
   '"Provisional 1 "',
   '"Not Towing"',
   '"Straight Ahead"',
   '"001"',
   '"5110"',
   '',
   '']),
 ('"SA"',
  ['"2015-2-21/08/2019"',
   '"02"',
   '1',
   '"SA"',
   '"Motor Cars - Sedan"',
   '"1994"',
   '"South East"',
   '"Female"',
   '"021"',
   '"C "',
   '"Full"',
   '"Not Towing"',
   '"Stopped on Carri

# 1.2
## Task 2B

In [12]:
#Task1.2
#2b

#A Function to print the data items in each RDD
def print_partitions(data):
    if isinstance(data, RDD):
        numPartitions = data.getNumPartitions()
        partitions = data.glom().collect()
    else:
        numPartitions = data.rdd.getNumPartitions()
        partitions = data.rdd.glom().collect()
    
    print(f"####### NUMBER OF PARTITIONS: {numPartitions}")
    for index, partition in enumerate(partitions):
        # show partition if it is not empty
        if len(partition) > 0:
            print(f"Partition {index}: {len(partition)} records")

# a method for hash partitioning
no_of_part=2
def hash_function(key):
    if key == '"SA"':
        return 0
    else:
        return 1



In [13]:
# partitioning the data using hash partitioning 
Partitioned_data=flatMapValues_unit_rdd_2.partitionBy(no_of_part,hash_function)


# 1.2
## Task 2C

In [14]:
#Task1.2
#2c

#A Function to print the data items in each RDD
def print_partitions(data):
    if isinstance(data, RDD):
        numPartitions = data.getNumPartitions()
        partitions = data.glom().collect()
    else:
        numPartitions = data.rdd.getNumPartitions()
        partitions = data.rdd.glom().collect()
    
    print(f"####### NUMBER OF PARTITIONS: {numPartitions}")
    for index, partition in enumerate(partitions):
        # show partition if it is not empty
        if len(partition) > 0:
            print(f"Partition {index}: {len(partition)} records")
            

In [15]:
#printing the partion records in each partion after hash partitioning
print_partitions(Partitioned_data)

####### NUMBER OF PARTITIONS: 2
Partition 0: 109684 records
Partition 1: 44170 records


$ \text {Since we have used the Hash partitioning strategy to partition the data on the bases of the LIC state we now have 2 partition on the unit_rdd RDD }\$

$\text {As we can see the results above can say the data is skewed and there are 109684 record for the SA state and just 
44170}\$ 

$ \text {for other states this signifies that the data is skewed and this will result in skewed processor as one processer will sit idle}\$ 


# 1.3
## Task 1

In [16]:
#Task 1.3
#1

unit_rdd2_male = unit_rdd1.filter(lambda x:x[8]!=0 and x[8] != '"XXX"' and x[7] == '"Male"')
unit_rdd3_male = unit_rdd2_male.map(lambda x:int(x[8][1:-1]))
#unit_rdd3 = unit_rdd2.filter(lambda x:(x[7] + int(x[8][1:-1])))
total_age_male = unit_rdd3_male.sum()
total_male_count = unit_rdd3_male.count()
average_male_age = total_age_male / total_male_count
round(average_male_age,2)

40.98

In [17]:
unit_rdd2_female = unit_rdd1.filter(lambda x:x[8]!=0 and x[8] != '"XXX"' and x[7] == '"Female"')
unit_rdd3_female = unit_rdd2_female.map(lambda x:int(x[8][1:-1]))
#unit_rdd3 = unit_rdd2.filter(lambda x:(x[7] + int(x[8][1:-1])))
total_age_female = unit_rdd3_female.sum()
total_female_count = unit_rdd3_female.count()
average_female_age = total_age_female / total_female_count
round(average_female_age,2)

40.39

## 1.3
## Task 2

In [18]:
# Task 1.3
#2
unit_rdd2 = unit_rdd1.filter(lambda x:(x[5] != '"XXXX"' and x[5] != ''))
unit_rdd3 = unit_rdd2.map(lambda x:(int(x[5][1:-1])))
unit_rdd4 = unit_rdd2.map(lambda x:((int(x[5][1:-1])),x[3] + " " + x[5] + " " + x[4]))
minYear = unit_rdd3.min()
maxYear = unit_rdd3.max()

In [19]:
unit_rdd_output_maxYear = unit_rdd4.filter(lambda x:x[0] == maxYear) 
unit_rdd_output_maxYear.collect()

[(2019, '"SA" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "OMNIBUS"'),
 (2019, '"SA" "2019" "Motor Cars - Sedan"'),
 (2019, '"SA" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "SEMI TRAILER"'),
 (2019, '"SA" "2019" "Motor Cars - Sedan"'),
 (2019, '"SA" "2019" "Motor Cars - Sedan"'),
 (2019, '"VIC" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "Utility"'),
 (2019, '"VIC" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "SEMI TRAILER"'),
 (2019, '"SA" "2019" "Utility"'),
 (2019, '"SA" "2019" "Utility"'),
 (2019, '"SA" "2019" "SEMI TRAILER"'),
 (2019, '"SA" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "Motor Cars - Sedan"'),
 (2019, '"SA" "2019" "Station Wagon"'),
 (2019, '"SA" "2019" "SEMI TRAILER"'),
 (2019, '"SA" "2019" "Utility"'),
 (2019, '"SA" "2019" "Utility"'),
 (2019, '"SA" "2019" "Utility"'),
 (2019, '"SA" "2019" "Other Defined Special Vehicle"'),
 (2019, '"SA" "2019" "Motor Cars - Sedan"'),
 (2

In [20]:
unit_rdd_output_minYear = unit_rdd4.filter(lambda x:x[0] == minYear)
unit_rdd_output_minYear.collect()

[(1900, '"VIC" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "RIGID TRUCK LGE GE 4.5T"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"'),
 (1900, '"SA" "1900" "Motor Cycle"')]

# 2.1
## Task 1

In [21]:
# Task 2.1
#1
#loading all the csv in a dataframe for crash and units
file_unit=['2015_DATA_SA_Units.csv','2016_DATA_SA_Units.csv','2017_DATA_SA_Units.csv',
           '2018_DATA_SA_Units.csv','2019_DATA_SA_Units.csv',]
df_unit = spark.read.format('csv')\
            .option('header',True)\
            .load(file_unit)
file_crash=['2015_DATA_SA_Crash.csv','2016_DATA_SA_Crash.csv','2017_DATA_SA_Crash.csv',
            '2018_DATA_SA_Crash.csv','2019_DATA_SA_Crash.csv',]
df_crash = spark.read.format('csv')\
            .option('header',True)\
            .load(file_crash)

# 2.1
## Task 2

In [22]:
#Task 2.1
#2
# printing the schema of unit and crash dataframe
df_unit.printSchema()
df_crash.printSchema()

root
 |-- REPORT_ID: string (nullable = true)
 |-- Unit No: string (nullable = true)
 |-- No Of Cas: string (nullable = true)
 |-- Veh Reg State: string (nullable = true)
 |-- Unit Type: string (nullable = true)
 |-- Veh Year: string (nullable = true)
 |-- Direction Of Travel: string (nullable = true)
 |-- Sex: string (nullable = true)
 |-- Age: string (nullable = true)
 |-- Lic State: string (nullable = true)
 |-- Licence Class: string (nullable = true)
 |-- Licence Type: string (nullable = true)
 |-- Towing: string (nullable = true)
 |-- Unit Movement: string (nullable = true)
 |-- Number Occupants: string (nullable = true)
 |-- Postcode: string (nullable = true)
 |-- Rollover: string (nullable = true)
 |-- Fire: string (nullable = true)

root
 |-- REPORT_ID: string (nullable = true)
 |-- Stats Area: string (nullable = true)
 |-- Suburb: string (nullable = true)
 |-- Postcode: string (nullable = true)
 |-- LGA Name: string (nullable = true)
 |-- Total Units: string (nullable = true)


# 2.2
## Task 1

In [23]:
#Task 2.2
#1
#filtering the data for adelaide and total casualty
df_count =df_crash.filter((df_crash['Suburb']=='ADELAIDE') & (df_crash['Total Cas'].cast('int') > 3))
df_count.show()

+--------------------+----------+--------+--------+----------------+-----------+---------+----------+--------+--------+----+--------+--------+--------+----------+-------------+----------------+--------------+--------------------+------------+-------------+------------+--------+--------------+---------+------------+-------------+---------------+------------+--------------+----------+----------+--------------+
|           REPORT_ID|Stats Area|  Suburb|Postcode|        LGA Name|Total Units|Total Cas|Total Fats|Total SI|Total MI|Year|   Month|     Day|    Time|Area Speed|Position Type|Horizontal Align|Vertical Align|          Other Feat|Road Surface|Moisture Cond|Weather Cond|DayNight|    Crash Type|Unit Resp| Entity Code|CSEF Severity|  Traffic Ctrls|DUI Involved|Drugs Involved|  ACCLOC_X|  ACCLOC_Y|    UNIQUE_LOC|
+--------------------+----------+--------+--------+----------------+-----------+---------+----------+--------+--------+----+--------+--------+--------+----------+-------------+

# 2.2
## Task 2

In [24]:
#Task 2.2
#2
df_sort = df_crash.withColumn('Total Cas',df_crash['Total Cas'].cast('int'))
df_sort=df_sort.select('*').sort('Total Cas', ascending=False).show(10)

+--------------------+--------------+---------------+--------+--------------------+-----------+---------+----------+--------+--------+----+--------+---------+--------+----------+-------------+--------------------+--------------+--------------+------------+-------------+------------+--------+-----------+---------+------------+-------------+---------------+------------+--------------+----------+----------+--------------+
|           REPORT_ID|    Stats Area|         Suburb|Postcode|            LGA Name|Total Units|Total Cas|Total Fats|Total SI|Total MI|Year|   Month|      Day|    Time|Area Speed|Position Type|    Horizontal Align|Vertical Align|    Other Feat|Road Surface|Moisture Cond|Weather Cond|DayNight| Crash Type|Unit Resp| Entity Code|CSEF Severity|  Traffic Ctrls|DUI Involved|Drugs Involved|  ACCLOC_X|  ACCLOC_Y|    UNIQUE_LOC|
+--------------------+--------------+---------------+--------+--------------------+-----------+---------+----------+--------+--------+----+--------+------

# 2.2
## Task 3

In [25]:
#Task 2.2
#3
df_count1 = df_crash.groupby(df_crash['Crash Type']).agg(f.sum(df_crash['Total Fats']).alias('Total Fatalities')).show()



+--------------------+----------------+
|          Crash Type|Total Fatalities|
+--------------------+----------------+
|           Roll Over|            57.0|
|  Hit Object on Road|             2.0|
|      Hit Pedestrian|            70.0|
|    Hit Fixed Object|           152.0|
|               Other|             2.0|
|          Side Swipe|            20.0|
|             Head On|            86.0|
|  Hit Parked Vehicle|             9.0|
|          Right Turn|            18.0|
|            Rear End|            16.0|
|          Hit Animal|             4.0|
|Left Road - Out o...|             1.0|
|         Right Angle|            45.0|
+--------------------+----------------+



# 2.2
## Task 4

In [26]:
#Task 2.2
#4
df_joinData= df_unit.join(df_crash,df_unit.REPORT_ID==df_crash.REPORT_ID,how='inner')
df_FatCount=df_joinData.filter(df_joinData['Licence Type']=='Unlicenced').groupby('Suburb')\
                .agg(f.sum(df_joinData['Total Cas']).alias('Total Casualties'))
df_FatCount.show()

+-------------------+----------------+
|             Suburb|Total Casualties|
+-------------------+----------------+
|      FLINDERS PARK|             8.0|
|       POOGINAGORIC|             1.0|
|     TEA TREE GULLY|             1.0|
|            HACKHAM|             3.0|
|   MEDINDIE GARDENS|             0.0|
|           WISANGER|             1.0|
|            CUMMINS|             0.0|
|       BASKET RANGE|             0.0|
|MURRAY BRIDGE SOUTH|             0.0|
|      GILLES PLAINS|             7.0|
|             HAWKER|             0.0|
|           BEAUFORT|             1.0|
|             MAGILL|             7.0|
|            ECHUNGA|             1.0|
|            CULTANA|             1.0|
|        EDWARDSTOWN|             6.0|
|        RISDON PARK|             1.0|
|          THORNGATE|             1.0|
|       ANDREWS FARM|            12.0|
|       TORRENSVILLE|             5.0|
+-------------------+----------------+
only showing top 20 rows



# 2.3
## Task 1

In [27]:
#Task 2.3
#1
df_SevCount=df_crash.groupby('CSEF Severity')\
                .agg(f.count(df_crash['Total Cas']).alias('Total Casualties'))
df_SevCount.show()

+-------------+----------------+
|CSEF Severity|Total Casualties|
+-------------+----------------+
|     4: Fatal|             451|
|        2: MI|           21881|
|       1: PDO|           46696|
|        3: SI|            2978|
+-------------+----------------+



# 2.3
## Task 2a

In [28]:
#Task 2.3
#2a
TotalDP=df_crash.filter(df_crash['Drugs Involved']=='Y').count()

df_SevCountDP=df_crash.filter(df_crash['Drugs Involved']=='Y').groupby('CSEF Severity')\
                .agg(f.count(df_crash['Total Cas']).alias('Count'))\
                .withColumn('Percent',f.round((f.col('Count')*100/TotalDP),2)).show()

+-------------+-----+-------+
|CSEF Severity|Count|Percent|
+-------------+-----+-------+
|     4: Fatal|   82|   6.54|
|        2: MI|  749|  59.73|
|       1: PDO|  176|  14.04|
|        3: SI|  247|   19.7|
+-------------+-----+-------+



# 2.3
## Task 2b

In [29]:
#Task 2.3
#2b
TotalBAC=df_crash.filter(df_crash['DUI Involved']=='Y').count()
df_SevCountBAC=df_crash.filter(df_crash['DUI Involved']=='Y').groupby('CSEF Severity')\
                .agg(f.count(df_crash['Total Cas']).alias('Count'))\
                .withColumn('Percent',f.round((f.col('Count')*100/TotalBAC),2)).show()

+-------------+-----+-------+
|CSEF Severity|Count|Percent|
+-------------+-----+-------+
|     4: Fatal|   79|   3.51|
|        2: MI|  737|  32.78|
|       1: PDO| 1173|  52.18|
|        3: SI|  259|  11.52|
+-------------+-----+-------+



# 2.3
## Task 2c

In [30]:
#Task 2.3
#2c
TotalDyAy=df_crash.filter((df_crash['Drugs Involved']=='Y')&(df_crash['DUI Involved']=='Y')).count()
df_SevCountDyAy=df_crash.filter((df_crash['Drugs Involved']=='Y')&(df_crash['DUI Involved']=='Y'))\
                .groupby('CSEF Severity').agg(f.count(df_crash['Total Cas']).alias('Count'))\
                .withColumn('Percent',f.round((f.col('Count')*100/TotalDyAy),2)).show()

+-------------+-----+-------+
|CSEF Severity|Count|Percent|
+-------------+-----+-------+
|     4: Fatal|   27|  15.43|
|        2: MI|   89|  50.86|
|       1: PDO|   24|  13.71|
|        3: SI|   35|   20.0|
+-------------+-----+-------+



# 2.3
## Task 2D

In [31]:
#Task 2.3
#2d
TotalDnAn=df_crash.filter((df_crash['Drugs Involved'].isNull())&(df_crash['DUI Involved'].isNull())).count()
df_SevCountDnAn=df_crash.filter((df_crash['Drugs Involved'].isNull())&(df_crash['DUI Involved'].isNull()))\
                        .groupby('CSEF Severity').agg(f.count(df_crash['Total Cas']).alias('Count'))\
                        .withColumn('Percent',f.round((f.col('Count')*100/TotalDnAn),2)).show()

+-------------+-----+-------+
|CSEF Severity|Count|Percent|
+-------------+-----+-------+
|     4: Fatal|  317|   0.46|
|        2: MI|20484|  29.83|
|       1: PDO|45371|  66.06|
|        3: SI| 2507|   3.65|
+-------------+-----+-------+



# 2.4
## Task A Dataframe

In [32]:
%%time
#Task 2.4
#a Dataframe
df_Left_join= df_unit.join(df_crash,df_unit.REPORT_ID==df_crash.REPORT_ID,how='left')
df_Left_join=df_Left_join.withColumn("Date",f.concat(f.col('Year'),f.lit('-'),\
                                                                 f.col('Month'),f.lit('-'),\
                                                                 f.col('Day')))
df_Left_join=df_Left_join.filter((df_Left_join['Suburb']=='ADELAIDE'))\
                .select('Date','Time','Total Cas','Sex','Age','Licence Type').show()

+--------------------+--------+---------+-------+----+------------+
|                Date|    Time|Total Cas|    Sex| Age|Licence Type|
+--------------------+--------+---------+-------+----+------------+
|2016-November-Wed...|01:45 pm|        1|   Male| 056|        Full|
|2016-November-Wed...|01:45 pm|        1|   Male| 072|        null|
|2016-November-Tue...|03:40 pm|        1|   Male| 056|        null|
|2016-November-Tue...|03:40 pm|        1| Female| 027|        null|
|2016-November-Tue...|05:00 pm|        0| Female| 032|        Full|
|2016-November-Tue...|05:00 pm|        0|Unknown| XXX|     Unknown|
|2016-November-Tue...|05:40 pm|        0|   Male| 022|     Unknown|
|2016-November-Tue...|05:40 pm|        0|   Male| 020|     Unknown|
|2016-November-Monday|11:26 pm|        0|Unknown| XXX|     Unknown|
|2016-November-Monday|11:26 pm|        0|   Male| 042|        Full|
|2016-November-Monday|11:26 pm|        0|   null|null|        null|
|2016-November-Monday|11:30 pm|        0|   Male

# 2.4
## Task A SQL

In [33]:
%%time
#Task 2.4
#a SQL
df_unit.createOrReplaceTempView("sql_unit")
df_crash.createOrReplaceTempView("sql_crash")

sql_unit_left_crash = spark.sql('''
  SELECT CONCAT(c.Year,'-' ,c.Month,'-' , c.Day) as Date , c.Time, c.`Total Cas`, u.Sex, u.Age, u.`Licence Type`
  FROM sql_unit u LEFT JOIN sql_crash c
  ON u.REPORT_ID=c.REPORT_ID
  Where c.Suburb = 'ADELAIDE'
''')

sql_unit_left_crash.show()
sql_unit_left_crash.count()

+--------------------+--------+---------+-------+----+------------+
|                Date|    Time|Total Cas|    Sex| Age|Licence Type|
+--------------------+--------+---------+-------+----+------------+
|2016-November-Wed...|01:45 pm|        1|   Male| 056|        Full|
|2016-November-Wed...|01:45 pm|        1|   Male| 072|        null|
|2016-November-Tue...|03:40 pm|        1|   Male| 056|        null|
|2016-November-Tue...|03:40 pm|        1| Female| 027|        null|
|2016-November-Tue...|05:00 pm|        0| Female| 032|        Full|
|2016-November-Tue...|05:00 pm|        0|Unknown| XXX|     Unknown|
|2016-November-Tue...|05:40 pm|        0|   Male| 022|     Unknown|
|2016-November-Tue...|05:40 pm|        0|   Male| 020|     Unknown|
|2016-November-Monday|11:26 pm|        0|Unknown| XXX|     Unknown|
|2016-November-Monday|11:26 pm|        0|   Male| 042|        Full|
|2016-November-Monday|11:26 pm|        0|   null|null|        null|
|2016-November-Monday|11:30 pm|        0|   Male

6310

# 2.4
## Task A RDD

In [34]:
%%time
#Task 2.4
#a RDD

flatMapValues_unit_rdd_2 = unit_rdd1.map(lambda x: (x[0], x[1:]))

flatMapValues_crash_rdd_2 = crash_rdd1.map(lambda x: (x[0], x[1:]))

rdd_join=flatMapValues_unit_rdd_2.join(flatMapValues_crash_rdd_2)

rdd_join1=rdd_join.filter(lambda x: x[1][1][1] == '"ADELAIDE"')
rdd_join1=rdd_join1.map(lambda x: (x[1][1][9]+'-'+x[1][1][10]+'-'+x[1][1][11],x[1][1][12],\
                                   x[1][1][5],x[1][0][6],x[1][0][7],x[1][0][10]))
rdd_join1.collect()

CPU times: user 24.2 ms, sys: 517 µs, total: 24.7 ms
Wall time: 3.48 s


[('2015-"January"-"Friday"', '"06:10 pm"', '1', '"Female"', '"035"', '"Full"'),
 ('2015-"January"-"Friday"', '"06:10 pm"', '1', '"Female"', '"037"', '"Full"'),
 ('2015-"January"-"Friday"', '"05:00 pm"', '0', '"Male"', '"041"', '"Full"'),
 ('2015-"January"-"Friday"', '"05:00 pm"', '0', '"Female"', '"022"', ''),
 ('2015-"February"-"Friday"',
  '"08:00 pm"',
  '1',
  '"Male"',
  '"020"',
  '"Provisional 1 "'),
 ('2015-"August"-"Monday"', '"09:07 am"', '0', '"Male"', '"039"', ''),
 ('2015-"August"-"Monday"', '"09:07 am"', '0', '"Male"', '"XXX"', '"Unknown"'),
 ('2015-"January"-"Monday"', '"01:15 pm"', '0', '"Male"', '"070"', '"Full"'),
 ('2015-"January"-"Monday"', '"01:15 pm"', '0', '"Female"', '"059"', '"Full"'),
 ('2015-"January"-"Wednesday"', '"08:15 pm"', '1', '"Male"', '"041"', ''),
 ('2015-"January"-"Tuesday"',
  '"03:10 pm"',
  '0',
  '"Female"',
  '"044"',
  '"Full"'),
 ('2015-"January"-"Tuesday"', '"03:10 pm"', '0', '"Male"', '"061"', '"Full"'),
 ('2015-"January"-"Wednesday"',
  '

# 2.4
## Task B Dataframe

In [35]:
%%time
#Task 2.4
#b Data frame
df_Left_join1= df_crash.join(df_unit,df_crash.REPORT_ID==df_unit.REPORT_ID,how='left')

df_Left_join1=df_Left_join1.filter(df_Left_join1['Licence Type']=='Unlicenced')\
                            .groupby(df_Left_join1['Suburb']).agg(f.sum(df_Left_join1['Total Cas'])\
                                                                  .alias('Total Casualties')).show()


+-------------------+----------------+
|             Suburb|Total Casualties|
+-------------------+----------------+
|      FLINDERS PARK|             8.0|
|       POOGINAGORIC|             1.0|
|     TEA TREE GULLY|             1.0|
|            HACKHAM|             3.0|
|   MEDINDIE GARDENS|             0.0|
|           WISANGER|             1.0|
|            CUMMINS|             0.0|
|       BASKET RANGE|             0.0|
|MURRAY BRIDGE SOUTH|             0.0|
|      GILLES PLAINS|             7.0|
|             HAWKER|             0.0|
|           BEAUFORT|             1.0|
|             MAGILL|             7.0|
|            ECHUNGA|             1.0|
|            CULTANA|             1.0|
|        EDWARDSTOWN|             6.0|
|        RISDON PARK|             1.0|
|          THORNGATE|             1.0|
|       ANDREWS FARM|            12.0|
|       TORRENSVILLE|             5.0|
+-------------------+----------------+
only showing top 20 rows

CPU times: user 3.99 ms, sys: 500 µs, 

# 2.4
## Task B SQL

In [36]:
%%time
#Task 2.4
#b SQL
df_unit.createOrReplaceTempView("sql_unit")
df_crash.createOrReplaceTempView("sql_crash")

sql_unit_left_crash1 = spark.sql('''
  SELECT c.Suburb , sum(c.`Total Cas`)
  FROM sql_crash c LEFT JOIN sql_unit u
  ON c.REPORT_ID=u.REPORT_ID
  Where u.`Licence Type` = 'Unlicenced'
  Group by c.Suburb
''')
# Displaying the output
sql_unit_left_crash1.show()

+-------------------+------------------------------+
|             Suburb|sum(CAST(Total Cas AS DOUBLE))|
+-------------------+------------------------------+
|      FLINDERS PARK|                           8.0|
|       POOGINAGORIC|                           1.0|
|     TEA TREE GULLY|                           1.0|
|            HACKHAM|                           3.0|
|   MEDINDIE GARDENS|                           0.0|
|           WISANGER|                           1.0|
|            CUMMINS|                           0.0|
|       BASKET RANGE|                           0.0|
|MURRAY BRIDGE SOUTH|                           0.0|
|      GILLES PLAINS|                           7.0|
|             HAWKER|                           0.0|
|           BEAUFORT|                           1.0|
|             MAGILL|                           7.0|
|            ECHUNGA|                           1.0|
|            CULTANA|                           1.0|
|        EDWARDSTOWN|                         

# 2.4
## Task B RDD

In [37]:
%%time
#Task 2.4
#b RDD
rdd_join2=rdd_join.filter(lambda x: (x[1][0][10]=='"Unlicenced"'))
rdd_join2=rdd_join2.map(lambda x: (x[1][1][1],int(x[1][1][5])))
rdd_final=rdd_join2.reduceByKey(lambda x,y: x+y)
rdd_final.collect()

CPU times: user 14.9 ms, sys: 198 µs, total: 15.1 ms
Wall time: 1.24 s


[('"MILE END"', 6),
 ('"SOUTH PLYMPTON"', 8),
 ('"TORRENSVILLE"', 5),
 ('"BIRKENHEAD"', 2),
 ('"ROSE PARK"', 0),
 ('"PORT AUGUSTA WEST"', 2),
 ('"MODBURY"', 2),
 ('"WINGFIELD"', 6),
 ('"SELLICKS HILL"', 1),
 ('"MENINGIE"', 2),
 ('"NORTHFIELD"', 2),
 ('"ELIZABETH DOWNS"', 6),
 ('"PORT LINCOLN"', 2),
 ('"BOWDEN"', 1),
 ('"SMITHFIELD"', 6),
 ('"GLENGOWRIE"', 1),
 ('"WHYALLA NORRIE"', 7),
 ('"GAWLER BELT"', 0),
 ('"TATACHILLA"', 1),
 ('"YAMBA"', 1),
 ('"ELIZABETH GROVE"', 9),
 ('"GOOLWA BEACH"', 1),
 ('"CURRENCY CREEK"', 5),
 ('"MEDINDIE GARDENS"', 0),
 ('"PARALOWIE"', 4),
 ('"MARION"', 4),
 ('"HAWKER"', 0),
 ('"BRIGHTON"', 2),
 ('"YALATA"', 1),
 ('"OAKLANDS PARK"', 3),
 ('"HILLIER"', 3),
 ('"BINNUM"', 1),
 ('"SEMAPHORE SOUTH"', 5),
 ('"BOLTO"', 0),
 ('"CRYSTAL BROOK"', 0),
 ('"OLARY"', 0),
 ('"CUNLIFFE"', 1),
 ('"BLAKEVIEW"', 3),
 ('"CHANDLERS HILL"', 1),
 ('"HACKHAM"', 3),
 ('"ENCOUNTER BAY"', 1),
 ('"RENMARK SOUTH"', 1),
 ('"ANANGU PITJANTJATJARA YANKUNYTJATJARA"', 0),
 ('"SEAFORD RISE"